File: examples/course-04-tool-mechanism/nodejs/src/agent.js
/**
 * 课程四工具机制示例的 Runtime 主循环。
 *
 * LLM 只负责提出结构化工具调用意图。Runtime 负责工具 Schema 暴露、参数校验、
 * 权限检查、审计、执行、Observation 处理和状态更新。
 */
const { SYSTEM_PROMPT } = require("./prompt");
const { executeToolCall } = require("./tools");

class AgentState {
  /**
   * Runtime 持有的单次任务状态。
   *
   * ``stepCount`` 只在工具调用步骤递增；final_answer / ask_user /
   * fail 等终止决策不计入步骤数，因此 trace 长度可能比 stepCount
   * 多 1（多出一条终止决策记录）。
   */
  constructor({ userGoal, maxSteps = 8, maxToolErrors = 5 }) {
    this.userGoal = userGoal;
    this.maxSteps = maxSteps;
    this.maxToolErrors = maxToolErrors;
    this.stepCount = 0;
    this.history = [];
    this.toolResults = [];
    this.errors = [];
    this.auditLog = [];
    this.stopReason = null;
    this.finalAnswer = null;
  }

  recentHistory(n = 5) {
    return this.history.slice(-n);
  }
}

async function runAgent({
  userGoal,
  registry,
  permissions,
  llmCall,
  systemPrompt = SYSTEM_PROMPT,
  maxSteps = 8,
  conversation = [],
  logger = null,
}) {
  // State 和 Trace 独立于 LLM 保存，这样循环行为可以被回放和审计。
  const state = new AgentState({ userGoal, maxSteps });
  const trace = [];

  // 0. 启动循环
  while (!state.stopReason) {
    // 1. Context Assembly：只把本轮决策需要的状态切片交给 LLM。
    const context = assembleContext({ systemPrompt, registry, state, conversation });

    // 2. 调用 llm 进行决策，并对结果进行标准化处理
    const decision = normalizeDecision(await llmCall(context));
    logEvent(logger, { event: "llm_decision", step: state.stepCount, decision });
    const traceEntry = {
      step: state.stepCount,
      contextSummary: contextSummary(context),
      decision,
    };

    // 3. 对结果进行判断
    if (decision.type === "final_answer") {
      state.stopReason = "completed";
      state.finalAnswer = decision.answer || "";
      return finalize({ state, traceEntry, trace, logger, status: "success", fields: { answer: state.finalAnswer } });
    }

    if (decision.type === "ask_user") {
      state.stopReason = "need_user_input";
      return finalize({ state, traceEntry, trace, logger, status: "paused", fields: { question: decision.question || "" } });
    }

    if (decision.type === "fail") {
      state.stopReason = "failed";
      return finalize({ state, traceEntry, trace, logger, status: "failed", fields: { reason: decision.reason || "" } });
    }

    // Interaction / Observation：Runtime 执行工具，并把结果统一成 Observation。
    logEvent(logger, {
      event: "tool_call",
      step: state.stepCount,
      toolName: decision.tool_name,
      arguments: decision.arguments || {},
    });

    // 4. 决策结果是要调用工具，执行工具
    const observation = await executeToolCall({ decision, registry, permissions, auditLog: state.auditLog });
    logEvent(logger, { event: "tool_result", step: state.stepCount, observation });

    // 5. 标准化处理工具执行结果并写入 state
    updateStateFromToolCall({ state, decision, observation });
    traceEntry.observation = observation;

    state.stepCount += 1;
    // 6. Continue or Stop：停止条件由 Runtime 判断，不依赖模型自觉。
    const stopReason = checkStop(state);
    if (stopReason) {
      state.stopReason = stopReason;
    }

    traceEntry.stateUpdate = stateSummary(state);
    traceEntry.stateSnapshot = stateSnapshot(state);
    traceEntry.stopCheck = { continue: !state.stopReason, reason: state.stopReason };
    logEvent(logger, { event: "state_update", step: state.stepCount, state: traceEntry.stateSnapshot });
    logEvent(logger, {
      event: "stop_check",
      step: state.stepCount,
      continue: !state.stopReason,
      reason: state.stopReason,
    });
    trace.push(traceEntry);
    // 7. 进行下一次循环
  }

  return { status: "stopped", reason: state.stopReason, state, trace };
}

function finalize({ state, traceEntry, trace, logger, status, fields = {} }) {
  /** 记录最终状态快照和停止检查，返回统一的结果结构。 */
  traceEntry.stateUpdate = stateSummary(state);
  traceEntry.stateSnapshot = stateSnapshot(state);
  traceEntry.stopCheck = { continue: false, reason: state.stopReason };
  logEvent(logger, { event: "state_update", step: state.stepCount, state: traceEntry.stateSnapshot });
  logEvent(logger, { event: "stop_check", step: state.stepCount, continue: false, reason: state.stopReason });
  trace.push(traceEntry);
  return { status, state, trace, ...fields };
}

function assembleContext({ systemPrompt, registry, state, conversation = [] }) {
  // 选择哪些状态切片应该进入下一次 LLM 调用。
  return {
    system: systemPrompt,
    goal: state.userGoal,
    conversation,
    history: state.recentHistory(5),
    tools: registry.toContextTools(),
    step: state.stepCount,
    maxSteps: state.maxSteps,
    errors: state.errors.slice(-3),
    toolResults: state.toolResults.slice(-5),
    auditLog: state.auditLog.slice(-5),
  };
}

function normalizeDecision(decision) {
  // 把模型输出收敛到 Runtime 能处理的决策格式。
  if (!decision || typeof decision !== "object" || Array.isArray(decision)) {
    return { type: "fail", thought: "Model output was not a JSON object.", reason: "invalid_decision" };
  }
  if (!["call_tool", "final_answer", "ask_user", "fail"].includes(decision.type)) {
    return { type: "fail", thought: "Model output had an unsupported type.", reason: "invalid_decision_type" };
  }
  if (decision.type === "call_tool" && !decision.arguments) {
    decision.arguments = {};
  }
  return decision;
}

function updateStateFromToolCall({ state, decision, observation }) {
  // 把本轮决策和 Observation 写回 State，供后续循环使用。
  state.history.push({ step: state.stepCount, decision, observation });
  state.toolResults.push({
    tool: decision.tool_name,
    status: observation.status,
    summary: observation.summary,
  });
  if (observation.status === "error") {
    state.errors.push(observation);
  }
}

function checkStop(state) {
  // 检查最大步数、错误次数和重复动作等硬停止条件。
  if (state.stepCount >= state.maxSteps) {
    return "max_steps_exceeded";
  }
  if (state.errors.length >= state.maxToolErrors) {
    return "tool_error_limit";
  }
  if (repeatedAction(state, 3)) {
    return "repeated_action";
  }
  return null;
}

function repeatedAction(state, threshold = 3) {
  // 检测 Agent 是否连续重复相同工具调用而没有推进。
  const recent = state.history.slice(-threshold);
  if (recent.length < threshold) {
    return false;
  }
  const signatures = recent.map((item) => {
    if (item.decision.type !== "call_tool") {
      return null;
    }
    return JSON.stringify({
      tool: item.decision.tool_name,
      arguments: sortObject(item.decision.arguments || {}),
    });
  });
  return signatures.every(Boolean) && new Set(signatures).size === 1;
}

function sortObject(value) {
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    return value;
  }
  return Object.keys(value)
    .sort()
    .reduce((acc, key) => {
      acc[key] = value[key];
      return acc;
    }, {});
}

function contextSummary(context) {
  return {
    goal: context.goal,
    conversationCount: context.conversation.length,
    step: context.step,
    historyCount: context.history.length,
    errorCount: context.errors.length,
    tools: context.tools.map((tool) => tool.name),
  };
}

function stateSummary(state) {
  return {
    stepCount: state.stepCount,
    historyCount: state.history.length,
    errorsCount: state.errors.length,
    auditCount: state.auditLog.length,
    stopReason: state.stopReason,
  };
}

function stateSnapshot(state) {
  // 生成当前步骤执行后的完整 State 快照，便于示例逐步打印。
  return {
    userGoal: state.userGoal,
    maxSteps: state.maxSteps,
    stepCount: state.stepCount,
    history: structuredClone(state.history),
    toolResults: structuredClone(state.toolResults),
    errors: structuredClone(state.errors),
    auditLog: structuredClone(state.auditLog),
    stopReason: state.stopReason,
    finalAnswer: state.finalAnswer,
  };
}

function logEvent(logger, event) {
  if (logger) {
    logger(event);
  }
}

module.exports = {
  AgentState,
  runAgent,
  assembleContext,
};

File: examples/course-04-tool-mechanism/python/tool_agent/agent.py
"""Runtime main loop for the course 04 tool mechanism example.

The LLM only proposes structured tool-call intent. The runtime owns tool Schema exposure, argument validation,
permission checks, auditing, execution, Observation processing, and state updates.
"""

from dataclasses import dataclass, field
from copy import deepcopy
from typing import Any, Callable, Dict, List, Optional

from tool_agent.prompt import SYSTEM_PROMPT
from tool_agent.tools import PermissionPolicy, ToolRegistry, execute_tool_call


Decision = Dict[str, Any]
Observation = Dict[str, Any]
LLMCall = Callable[[Dict[str, Any]], Decision]
Logger = Callable[[Dict[str, Any]], None]


@dataclass
class AgentState:
    """Single-task state held by the runtime.

    ``step_count`` increments only on tool-call steps; final_answer / ask_user /
    fail and other terminal decisions do not count as steps, so trace length may be
    one greater than step_count because it includes the terminal decision record.
    """

    user_goal: str
    max_steps: int = 8
    step_count: int = 0
    history: List[Dict[str, Any]] = field(default_factory=list)
    tool_results: List[Dict[str, Any]] = field(default_factory=list)
    errors: List[Observation] = field(default_factory=list)
    audit_log: List[Dict[str, Any]] = field(default_factory=list)
    stop_reason: Optional[str] = None
    final_answer: Optional[str] = None
    max_tool_errors: int = 5

    def recent_history(self, n: int = 5) -> List[Dict[str, Any]]:
        return self.history[-n:]


def run_agent(
    user_goal: str,
    registry: ToolRegistry,
    permissions: PermissionPolicy,
    llm_call: LLMCall,
    system_prompt: str = SYSTEM_PROMPT,
    max_steps: int = 8,
    conversation: Optional[List[Dict[str, str]]] = None,
    logger: Optional[Logger] = None,
) -> Dict[str, Any]:
    """Run a ReAct-style loop until the task is complete, paused, failed, or stopped by hard constraints."""
    state = AgentState(user_goal=user_goal, max_steps=max_steps)
    trace: List[Dict[str, Any]] = []

    # 0. Start the loop
    while not state.stop_reason:
        # 1. Context Assembly: pass only the state slices needed for this decision to the LLM.
        context = assemble_context(system_prompt, registry, state, conversation=conversation or [])
        # 2. Call the LLM for a decision and normalize the output
        decision = _normalize_decision(llm_call(context))
        _log(logger, {"event": "llm_decision", "step": state.step_count, "decision": decision})
        trace_entry = {"step": state.step_count, "context_summary": _context_summary(context), "decision": decision}

        # 3. Inspect the decision result and decide whether to stop the loop
        decision_type = decision["type"]
        if decision_type == "final_answer":
            state.stop_reason = "completed"
            state.final_answer = decision.get("answer", "")
            return _finalize(state, trace_entry, trace, logger, "success", answer=state.final_answer)

        if decision_type == "ask_user":
            state.stop_reason = "need_user_input"
            return _finalize(state, trace_entry, trace, logger, "paused", question=decision.get("question", ""))

        if decision_type == "fail":
            state.stop_reason = "failed"
            return _finalize(state, trace_entry, trace, logger, "failed", reason=decision.get("reason", ""))

        # 4. If the decision requires a tool call, execute it
        # Interaction / Observation: the runtime executes tools and normalizes results into Observations.
        _log(
            logger,
            {
                "event": "tool_call",
                "step": state.step_count,
                "tool_name": decision.get("tool_name"),
                "arguments": decision.get("arguments", {}),
            },
        )
        observation = execute_tool_call(
            decision,
            registry=registry,
            permissions=permissions,
            audit_log=state.audit_log,
        )
        _log(logger, {"event": "tool_result", "step": state.step_count, "observation": observation})

        # 5. Normalize the tool-call result and write it into state
        _update_state_from_tool_call(state, decision, observation)
        trace_entry["observation"] = observation

        # 6. Update state and decide whether to enter the next loop iteration
        state.step_count += 1
        # Continue or Stop: the runtime owns stopping conditions instead of relying on the model to stop itself.
        stop_reason = _check_stop(state)
        if stop_reason:
            state.stop_reason = stop_reason

        trace_entry["state_update"] = _state_summary(state)
        trace_entry["state_snapshot"] = _state_snapshot(state)
        trace_entry["stop_check"] = {"continue": not state.stop_reason, "reason": state.stop_reason}
        _log(logger, {"event": "state_update", "step": state.step_count, "state": trace_entry["state_snapshot"]})
        _log(
            logger,
            {
                "event": "stop_check",
                "step": state.step_count,
                "continue": not state.stop_reason,
                "reason": state.stop_reason,
            },
        )
        trace.append(trace_entry)
        # 7. Continue to the next loop iteration

    return {"status": "stopped", "reason": state.stop_reason, "state": state, "trace": trace}


def assemble_context(
    system_prompt: str,
    registry: ToolRegistry,
    state: AgentState,
    conversation: Optional[List[Dict[str, str]]] = None,
) -> Dict[str, Any]:
    """Choose which state slices should enter the next LLM call."""
    return {
        "system": system_prompt,
        "goal": state.user_goal,
        "conversation": conversation or [],
        "history": state.recent_history(n=5),
        "tools": registry.to_context_tools(),
        "step": state.step_count,
        "max_steps": state.max_steps,
        "errors": state.errors[-3:],
        "tool_results": state.tool_results[-5:],
        "audit_log": state.audit_log[-5:],
    }


def _normalize_decision(decision: Decision) -> Decision:
    """Converge model output into a decision format the runtime can handle."""
    if not isinstance(decision, dict):
        return {"type": "fail", "thought": "Model output was not a JSON object.", "reason": "invalid_decision"}
    if decision.get("type") not in ("call_tool", "final_answer", "ask_user", "fail"):
        return {"type": "fail", "thought": "Model output had an unsupported type.", "reason": "invalid_decision_type"}
    if decision["type"] == "call_tool":
        decision.setdefault("arguments", {})
    return decision


def _update_state_from_tool_call(state: AgentState, decision: Decision, observation: Observation) -> None:
    """Write this turn's decision and Observation back to State for later loop iterations."""
    state.history.append({"step": state.step_count, "decision": decision, "observation": observation})
    state.tool_results.append(
        {"tool": decision.get("tool_name"), "status": observation["status"], "summary": observation["summary"]}
    )
    if observation["status"] == "error":
        state.errors.append(observation)


def _finalize(
    state: AgentState,
    trace_entry: Dict[str, Any],
    trace: List[Dict[str, Any]],
    logger: Optional[Logger],
    status: str,
    **kwargs: Any,
) -> Dict[str, Any]:
    """Record the final state snapshot and stop check, then return a normalized result structure."""
    trace_entry["state_update"] = _state_summary(state)
    trace_entry["state_snapshot"] = _state_snapshot(state)
    trace_entry["stop_check"] = {"continue": False, "reason": state.stop_reason}
    _log(logger, {"event": "state_update", "step": state.step_count, "state": trace_entry["state_snapshot"]})
    _log(logger, {"event": "stop_check", "step": state.step_count, "continue": False, "reason": state.stop_reason})
    trace.append(trace_entry)
    return {"status": status, "state": state, "trace": trace, **kwargs}


def _check_stop(state: AgentState) -> Optional[str]:
    """Check hard stopping conditions such as maximum steps, error count, and repeated actions."""
    if state.step_count >= state.max_steps:
        return "max_steps_exceeded"
    if len(state.errors) >= state.max_tool_errors:
        return "tool_error_limit"
    if _repeated_action(state, threshold=3):
        return "repeated_action"
    return None


def _repeated_action(state: AgentState, threshold: int = 3) -> bool:
    """Detect whether the agent repeats the same tool call without making progress."""
    recent = state.history[-threshold:]
    if len(recent) < threshold:
        return False
    signatures = []
    for item in recent:
        decision = item["decision"]
        if decision.get("type") != "call_tool":
            return False
        signatures.append((decision.get("tool_name"), repr(sorted(decision.get("arguments", {}).items()))))
    return len(set(signatures)) == 1


def _context_summary(context: Dict[str, Any]) -> Dict[str, Any]:
    return {
        "goal": context["goal"],
        "conversation_count": len(context["conversation"]),
        "step": context["step"],
        "history_count": len(context["history"]),
        "error_count": len(context["errors"]),
        "tools": [tool["name"] for tool in context["tools"]],
    }


def _state_summary(state: AgentState) -> Dict[str, Any]:
    return {
        "step_count": state.step_count,
        "history_count": len(state.history),
        "errors_count": len(state.errors),
        "audit_count": len(state.audit_log),
        "stop_reason": state.stop_reason,
    }


def _state_snapshot(state: AgentState) -> Dict[str, Any]:
    """Build a full State snapshot after the current step so the example can print each step."""
    return {
        "user_goal": state.user_goal,
        "max_steps": state.max_steps,
        "step_count": state.step_count,
        "history": deepcopy(state.history),
        "tool_results": deepcopy(state.tool_results),
        "errors": deepcopy(state.errors),
        "audit_log": deepcopy(state.audit_log),
        "stop_reason": state.stop_reason,
        "final_answer": state.final_answer,
    }


def _log(logger: Optional[Logger], event: Dict[str, Any]) -> None:
    if logger is not None:
        logger(event)
